### ISSUE TITLE
Determining Transaction Status

### ISSUE THREAD

**Issue by taspotts** [awslabs/pgbouncer-rr-patch#44]
>Multi-statement transactions will work correctly only if statements are not rerouted by the routing_rules function to a different server pool mid-transaction.

Would comparing xact_start to query_start be a reliable method to determine if the client is in transaction, or is there some other attribute we can examine, then pass this as a argument to the routing function?  Trying to prevent multi-statement transactions from getting rerouted incorrectly.

From `client.c`:
```
        /* remember timestamp of the first query in a transaction */
	if (!client->xact_start) {
		client->pool->stats.xact_count++;
		client->xact_start = client->query_start;
	}
```

Note how xact_start is `1573580794947376` on all these commands:
```
$ tail -f /tmp/pgbouncer.log | grep 'route_client_connection'
2019-11-12 17:46:34.947 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Username => username
2019-11-12 17:46:34.947 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Query => BEGIN;
2019-11-12 17:46:34.947 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: xact_start => 1573580794947376
2019-11-12 17:46:34.947 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: query_start => 1573580794947376
2019-11-12 17:46:34.948 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Username => username
2019-11-12 17:46:34.948 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Query => BEGIN;
2019-11-12 17:46:34.948 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: xact_start => 1573580794947376
2019-11-12 17:46:34.948 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: query_start => 1573580794947376
2019-11-12 17:46:35.041 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Username => username
2019-11-12 17:46:35.041 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Query => SELECT * FROM pgbench_branches LIMIT 11;
2019-11-12 17:46:35.041 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: xact_start => 1573580794947376
2019-11-12 17:46:35.041 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: query_start => 1573580795041918
2019-11-12 17:46:35.124 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Username => username
2019-11-12 17:46:35.124 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Query => UPDATE pgbench_branches SET bid = bid WHERE bid < 200;
2019-11-12 17:46:35.124 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: xact_start => 1573580794947376
2019-11-12 17:46:35.124 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: query_start => 1573580795124458
2019-11-12 17:46:35.219 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Username => username
2019-11-12 17:46:35.219 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Query => SELECT * FROM pgbench_branches LIMIT 7;
2019-11-12 17:46:35.219 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: xact_start => 1573580794947376
2019-11-12 17:46:35.219 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: query_start => 1573580795219278
2019-11-12 17:46:35.316 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Username => username
2019-11-12 17:46:35.316 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: Query => COMMIT;
2019-11-12 17:46:35.316 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: xact_start => 1573580794947376
2019-11-12 17:46:35.316 9587 DEBUG C-0x55dee77e8950: postgres/username@127.0.0.1:59510 route_client_connection: query_start => 1573580795316069
```

**Comment by rstrahan** [comment]
That looks like a very promising approach @taspotts - thanks!  It would be great if you want to implement and test this. Pass xact_start to the routing/rewrite functions as an additional arg.  If it works, feel free to send a PR.

**Comment by taspotts** [comment]
C programming is not my forte, so anything I can figure out would be an awful hack at best...  But I can try!  :)

Another thought on this per:  https://www.pgcon.org/2014/schedule/attachments/330_postgres-for-the-wire.pdf

ReadyForQuery - the ReadyForQuery message includes transaction status
'Z' int32 len 'I' or 'T' or 'E'

Current backend transaction status indicator. Possible values are ‘I’ if idle (not in a transaction block); ‘T’ if in a transaction block; or ‘E’ if in a failed transaction block (queries will be rejected until block is ended).

Is this information accessible from the client/pool/link "object"?  I don't even know where to begin looking.  :(

**Comment by ibot3** [comment]
pgbouncer puts the transaction status in `client->link->idle_tx`. This value is `true`/`1` if the state is either `T` or `E`.

I made some modification, so that the parameter is used in the python call: https://github.com/ibot3/pgbouncer-rr-patch/pull/1

However, these changes currently depend on #53. (I also made a python3.7 branch on my fork).

**Comment by n-oden** [comment]
@ibot3 #53 is now merged.  I'd love to see https://github.com/ibot3/pgbouncer-rr-patch/pull/1 (and possibly the other various fixes from https://github.com/athenianco/pgbouncer-rr-patch) merged into here as well, and I'm very curious how you ended up using that change in production -- the idle_tx boolean isn't set until after the transaction is started (so for example `BEGIN` on a single line doesn't have it), and detecting that seems like it's littered with edge cases.

**Comment by ibot3** [comment]
I currently do not have time or motivation to create a new pull request. However, it's just a few changes that are probably made quickly.

I did some advanced tests with this setup but it did not ended in production. However it may could have worked.
I created a (kind of hacky) python routing script that checks with a regex and the transaction status if the query should be routed to the primary or to the replication servers. I also added a healthcheck, so that queries are only sent to healthy servers.

https://gist.github.com/ibot3/47dcd35065f2aaa15562aae03d6a53cb
Feel free to use it, but I don't know how it would perform in production and with a higher load.

**Comment by mugunthan7285** [comment]
Actually without [ibot3#1](https://github.com/ibot3/pgbouncer-rr-patch/pull/1) I am facing problem with normal select queries also.
**With [ibot3#1](https://github.com/ibot3/pgbouncer-rr-patch/pull/1)**
```
2024-10-21 22:48:27.871  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:53926 route_client_connection: Username => root
2024-10-21 22:48:27.871  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:53926 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:48:27.871  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:53926 route_client_connection: xact_start => 1729576107871595
2024-10-21 22:48:27.871  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:53926 route_client_connection: query_start => 1729576107871595
2024-10-21 22:48:27.871  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:53926 route_client_connection: in_transaction => 0
2024-10-21 22:48:27.888  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection: Username => root
2024-10-21 22:48:27.888  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:48:27.888  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection: xact_start => 1729576107871595
2024-10-21 22:48:27.888  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection: query_start => 1729576107871595
2024-10-21 22:48:27.888  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection: in_transaction => 0
2024-10-21 22:48:27.890  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection: Username => root
2024-10-21 22:48:27.890  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:48:27.890  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection: xact_start => 1729576107871595
2024-10-21 22:48:27.890  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection: query_start => 1729576107871595
2024-10-21 22:48:27.890  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection:in_transaction value updated to 0 using client->link->idle_tx
2024-10-21 22:48:27.890  [2771888] DEBUG C-0x55696541d640: postgres_read_db/root@[::1]:53926 route_client_connection: in_transaction => 0
2024-10-21 22:48:28.623  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:37160 route_client_connection: Username => root
2024-10-21 22:48:28.623  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:37160 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:48:28.623  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:37160 route_client_connection: xact_start => 1729576108623006
2024-10-21 22:48:28.623  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:37160 route_client_connection: query_start => 1729576108623006
2024-10-21 22:48:28.623  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:37160 route_client_connection: in_transaction => 0
2024-10-21 22:48:29.126  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:37162 route_client_connection: Username => root
2024-10-21 22:48:29.126  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:37162 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:48:29.126  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:37162 route_client_connection: xact_start => 1729576109126405
2024-10-21 22:48:29.126  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:37162 route_client_connection: query_start => 1729576109126405
2024-10-21 22:48:29.126  [2771888] DEBUG C-0x55696541d640: postgres/root@[::1]:37162 route_client_connection: in_transaction => 0
```
**Without [ibot3#1](https://github.com/ibot3/pgbouncer-rr-patch/pull/1)**
```
2024-10-21 22:50:14.800  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43342 route_client_connection: Username => root
2024-10-21 22:50:14.800  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43342 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:50:14.800  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43342 route_client_connection: xact_start => 1729576214800673
2024-10-21 22:50:14.800  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43342 route_client_connection: query_start => 1729576214800673
2024-10-21 22:50:14.800  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43342 route_client_connection: in_transaction => 0
2024-10-21 22:50:14.817  [2775878] DEBUG C-0x5640fecba640: postgres_read_db/root@[::1]:43342 route_client_connection: Username => root
2024-10-21 22:50:14.817  [2775878] DEBUG C-0x5640fecba640: postgres_read_db/root@[::1]:43342 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:50:14.817  [2775878] DEBUG C-0x5640fecba640: postgres_read_db/root@[::1]:43342 route_client_connection: xact_start => 1729576214800673
2024-10-21 22:50:14.818  [2775878] DEBUG C-0x5640fecba640: postgres_read_db/root@[::1]:43342 route_client_connection: query_start => 1729576214800673
2024-10-21 22:50:14.818  [2775878] DEBUG C-0x5640fecba640: postgres_read_db/root@[::1]:43342 route_client_connection: in_transaction => 1
2024-10-21 22:50:14.821  [2775878] DEBUG C-0x5640fecba640: postgres_write_db/root@[::1]:43342 route_client_connection: Username => root
2024-10-21 22:50:14.821  [2775878] DEBUG C-0x5640fecba640: postgres_write_db/root@[::1]:43342 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:50:14.821  [2775878] DEBUG C-0x5640fecba640: postgres_write_db/root@[::1]:43342 route_client_connection: xact_start => 1729576214800673
2024-10-21 22:50:14.821  [2775878] DEBUG C-0x5640fecba640: postgres_write_db/root@[::1]:43342 route_client_connection: query_start => 1729576214800673
2024-10-21 22:50:14.822  [2775878] DEBUG C-0x5640fecba640: postgres_write_db/root@[::1]:43342 route_client_connection: in_transaction => 1
2024-10-21 22:50:14.824  [2775878] DEBUG C-0x5640fecba640: postgres_write_db/root@[::1]:43342 route_client_connection: Username => root
2024-10-21 22:50:14.824  [2775878] DEBUG C-0x5640fecba640: postgres_write_db/root@[::1]:43342 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:50:14.824  [2775878] DEBUG C-0x5640fecba640: postgres_write_db/root@[::1]:43342 route_client_connection: xact_start => 1729576214800673
2024-10-21 22:50:14.824  [2775878] DEBUG C-0x5640fecba640: postgres_write_db/root@[::1]:43342 route_client_connection: query_start => 1729576214800673
2024-10-21 22:50:14.824  [2775878] DEBUG C-0x5640fecba640: postgres_write_db/root@[::1]:43342 route_client_connection: in_transaction => 1
2024-10-21 22:50:15.412  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43348 route_client_connection: Username => root
2024-10-21 22:50:15.412  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43348 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:50:15.412  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43348 route_client_connection: xact_start => 1729576215412572
2024-10-21 22:50:15.412  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43348 route_client_connection: query_start => 1729576215412572
2024-10-21 22:50:15.412  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43348 route_client_connection: in_transaction => 0
2024-10-21 22:50:15.416  [2775878] DEBUG C-0x5640fecba640: postgres_read_db/root@[::1]:43348 route_client_connection: Username => root
2024-10-21 22:50:15.416  [2775878] DEBUG C-0x5640fecba640: postgres_read_db/root@[::1]:43348 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:50:15.416  [2775878] DEBUG C-0x5640fecba640: postgres_read_db/root@[::1]:43348 route_client_connection: xact_start => 1729576215412572
2024-10-21 22:50:15.416  [2775878] DEBUG C-0x5640fecba640: postgres_read_db/root@[::1]:43348 route_client_connection: query_start => 1729576215412572
2024-10-21 22:50:15.416  [2775878] DEBUG C-0x5640fecba640: postgres_read_db/root@[::1]:43348 route_client_connection: in_transaction => 1
2024-10-21 22:50:15.907  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43364 route_client_connection: Username => root
2024-10-21 22:50:15.907  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43364 route_client_connection: Query => select inet_server_addr()
2024-10-21 22:50:15.907  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43364 route_client_connection: xact_start => 1729576215907496
2024-10-21 22:50:15.907  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43364 route_client_connection: query_start => 1729576215907496
2024-10-21 22:50:15.907  [2775878] DEBUG C-0x5640fecba640: postgres/root@[::1]:43364 route_client_connection: in_transaction => 0

```
I have executed simple select query **select inet_server_addr()**
See that in second log, in_tranaction value is  1, eventhough it is not an explicit transaction.
I am using **transaction** pool_mode btw. In transaction pool_mode, **first 2 queries are redirected to postgres_write_db** , and the **3 rd query is redirected to postgres_read_db** . While in **session** pool_mode , **all queries are going to postgres_write_db** only.

**routing_rules.py**
```
def routing_rules(username, query, in_transaction):
	if ("select" in query or "SELECT" in query) and not in_transaction:
		return "postgres_read_db"
	else:
		return "postgres_write_db"
```
I have attached the whole log files here.
[bouncer.bugfix.log](https://github.com/user-attachments/files/17471300/bouncer.bugfix.log)
[bouncer.def.log](https://github.com/user-attachments/files/17471314/bouncer.def.log)

Let me know if more information needed :)
